compiler/src/main/kotlin/io/spine/protodata/codegen/BaseTypeSystem.kt

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2023, TeamDev. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.protodata.codegen
+
+import io.spine.protodata.PrimitiveType
+import io.spine.protodata.Type
+import io.spine.protodata.TypeName
+import io.spine.protodata.Value
+import io.spine.protodata.Value.KindCase.BOOL_VALUE
+import io.spine.protodata.Value.KindCase.BYTES_VALUE
+import io.spine.protodata.Value.KindCase.DOUBLE_VALUE
+import io.spine.protodata.Value.KindCase.ENUM_VALUE
+import io.spine.protodata.Value.KindCase.INT_VALUE
+import io.spine.protodata.Value.KindCase.LIST_VALUE
+import io.spine.protodata.Value.KindCase.MAP_VALUE
+import io.spine.protodata.Value.KindCase.MESSAGE_VALUE
+import io.spine.protodata.Value.KindCase.NULL_VALUE
+import io.spine.protodata.Value.KindCase.STRING_VALUE
+
+/**
+ * The type system of an application being built by ProtoData.
+ *
+ * The type system knows the association between the Protobuf types and the types of
+ * a target language.
+ *
+ * @param T the type of a language-specific type name
+ * @param V the type of a language-specific expression, i.e. the code that yields a value
```

Comment:
Now we separate this knowledge into two generic types, while they should correspond to each other. For example, I should not be able to match an `int` from C to `Integer` from Pascal. While currently these generics allow doing so.

The match between the types is and expression is a thing which should reside somewhere near `Language`.